[2단계 - 출석 구현] 랜디(김정우) 미션 제출합니다.

KJungW opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3906


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

코드의 가독성이나 유지보수성을 해치지 않는 선에서 객체지향 생활체조 원칙을 일정 수준으로 준수했다고 생각합니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### 질문 리스트

- TDD 진행방식에 질문이 있습니다.
  - 이번에 가장 큰 도메인인 AttendanceSystem을 기준으로 설계없이 TDD를 진행했습니다. 진행하면서 AttendanceSystem의 내부 도메인들은 언제든지 바뀔 수 있다는 생각에 AttendanceSystem에 대한 테스트만을 작성하며 TDD로 기능구현을 진행해나갔고, 어느정도 기능구현이 완료되었을 때 AttendanceSystem 내부 도메인들을 확정시키고 해당 내부 도메인들에 대한 테스트를 작성했습니다. TDD는 항상 실패 테스트 먼저라고 배우기는 했는데 내부 도메인들이나 설계가 확정되지 않은 상태에서 내부 도메인에 대한 테스트를 작성하는 것은 비효율적이라고 생각해 이러한 방식으로 진행했고, 이전 TDD를 진행할때보다 TDD의 이점이 잘 느껴졌습니다. 현재 진행방식이 바람직한 방식인지 조앤의 의견이 궁금합니다!

 - 커스텀 예외 생성 기준이 궁금합니다.
    - 커스텀 예외를 통해 예외의 범위를 나누라는 것을 배우고 이번 미션에서는 커스텀 예외를 만들어 적용했습니다. 항상 할때마다 커스텀 예외를 어떤 기준으로 나눠야할지 고민이 많이 되는데 조앤은 어떤 방식으로 커스텀 예외를 생성하시는지 궁금합니다!

### 1단계 미션과 피드백을 통해 새롭게 배운점 정리
- 요구사항 정리방법 정립
- TDD를 다운업 방식이 아닌 탑다운 방식으로 진행하는 방법
- 인텔리제이 단축키 학습 및 정리
- LocalDateTime/LocalDate/LocalTime 관련 기능 학습
- DateTimeFormatter를 활용한 LocalDateTime/LocalDate/LocalTime 파싱과 포매팅
- record를 활용한 DTO 간편 생성
- 도메인은 뷰를 의존하면 안되지만 뷰는 도메인을 어느정도 의존해도 괜찮다.
- 컨벤션은 코드의 가독성을 위해 존재하는 것이다. 컨벤션을 무리하게 적용해서 코드의 가독성을 떨어뜨리면 X

KJungW committed:
init(application): 2단계 미션을 위한 초기화

작업 내용
- 2단계 미션 진행을 위하 커밋 초기화

KJungW committed:
docs(readme): 요구사항 정리

작업내용
- 애플리케이션 요구사항을 README.md에 정리

KJungW committed:
test(add_record): 출석 추가 (실패 테스트)

테스트 내용
- 닉네임과 출석 시간으로 출석 기록을 추가할 수 있다

KJungW committed:
feat(add_record): 출석 추가 (구현)

작업 내용
- 닉네임과 출석 시간으로 출석 기록을 추가할 수 있다

KJungW committed:
test(add_record): 출석 상태 결정 (실패 테스트)

테스트 내용
- 교육시간과 출석정책을 기준으로 월요일의 출석 상태를 결정한다
- 교육시간과 출석정책을 기준으로 화요일에서 금요일의 출석 상태를 결정한다

KJungW committed:
feat(add_record): 출석 상태 결정 (구현)

작업 내용
- 교육시간과 출석정책을 기준으로 월요일의 출석 상태를 결정한다
- 교육시간과 출석정책을 기준으로 화요일에서 금요일의 출석 상태를 결정한다

KJungW committed:
test(add_record): 이미 등교한 경우 처리 (실패 테스트)

테스트 내용
- 이미 출석한 경우, 다시 출석할 수 없으며 수정 기능을 이용하도록 안내한다

KJungW committed:
feat(add_record): 이미 등교한 경우 처리 (구현)

작업 내용
- 이미 출석한 경우, 다시 출석할 수 없으며 수정 기능을 이용하도록 안내한다

KJungW committed:
feat(add_record): 등록되지 않은 크루 처리 (실패 테스트 작성)

작업 내용
- 네임이 등록되지 않은 경우 예외 메세지를 출력한다

KJungW committed:
feat(add_record): 등록되지 않은 크루 처리 (구현)

구현내용
- 네임이 등록되지 않은 경우 예외 메세지를 출력한다

KJungW committed:
test(add_record): 등교일 체크 (실패 테스트 구현)

작업 내용
- 등교일이 아닌 경우 예외 메세지를 출력한다

KJungW committed:
feat(add_record): 등교일 체크 (구현)

구현내용
- 등교일이 아닌 경우 예외 메세지를 출력한다

KJungW committed:
test(add_record): 캠퍼스 운영시간 체크 (실패 테스트 구현)

테스트 내용
- 캠퍼스 운영 시간이 아닌 경우 예외 메세지를 출력한다
- 캠퍼스 운영 시간인 경우 예외 메세지를 발생시키지 않는다

KJungW committed:
feat(add_record): 캠퍼스 운영시간 체크 (구현)

구현 내용
- 캠퍼스 운영 시간이 아닌 경우 예외 메세지를 출력한다
- 캠퍼스 운영 시간인 경우 예외 메세지를 발생시키지 않는다

KJungW committed:
refactor(add_record): 클래스 기능 분배

작업내용
- 클래스에게 각자의 역할에 맞게 기능 분배
- 도메인을 주제별로 나눠 패키지로 분리

KJungW committed:
test(add_record): 테스트명 정리

작업 내용
- 다른 주요 기능과 테스트명이 겹치지 않도록
  테스트명에 컨벤션 추가

KJungW committed:
test(update_record): 출석 기록 수정 (실패하는 테스트 작성)

작업내용
- 닉네임, 수정 목표 날짜, 새로운 출석 시간으로 기존 출석 기록을 수정할 수 있다

KJungW committed:
feat(update_record): 출석 기록 수정 (구현)

작업내용
- 출석 기록 수정 - 닉네임, 수정 목표 날짜,
  새로운 출석 시간으로 기존 출석 기록을 수정할 수 있다

KJungW committed:
test(update_record): 출석 상태 계산 (실패하는 테스트)

테스트 내용
- 교육시간과 출석정책을 기준으로 출석 상태를 결정한다

KJungW committed:
test(update_record): 등록되지 않은 크루 처리 (실패 테스트 작성)

테스트 내용
- 닉네임이 등록되지 않은 경우 예외 메세지를 출력한다

KJungW committed:
feat(update_record): 등록되지 않은 크루 처리 (구현)

작업내용
- 닉네임이 등록되지 않은 경우 예외 메세지를 출력한다

KJungW committed:
test(update_record): 등교일이 아닌 경우 처리 (실패 테스트 작성)

작업 내용
- 등교일이 아닌 경우 예외 메세지를 출력한다

KJungW committed:
test(update_record): 캠퍼스 시간이 이닌 경우 처리 (실패 테스트 작성)

테스트 내용
- 캠퍼스 운영 시간이 아닌 경우 예외 메세지를 출력한다

KJungW committed:
test(update_record): 출석 조회 (실패 테스트 작성)

테스트내용
- 닉네임을 통해 해당 크루의 출석 기록 일자순으로 조회할 수 있다

KJungW committed:
test(update_record): 테스트 코드 간소화

작업내용
- 암시적으로 초기화되는 부분은 beforeEach()에서 제거

KJungW committed:
test(update_record): 잘못된 실패 테스트 수정

작업내용
- 잘못된 실패테스트 내용 수정

KJungW committed:
feat(search_record): 출석 조회 (구현)

구현내용
- 닉네임을 통해 해당 크루의 출석 기록 일자순으로 조회할 수 있다

KJungW committed:
feat(update_record): 출석 현황 조회 (실패 테스트 작성)

테스트 내용
- 닉네임을 통해 해당 크루의 출석 현황을 조회할 수 있다

KJungW committed:
feat(update_record): 출석 현황 조회(구현)

구현내용
- 닉네임을 통해 해당 크루의 출석 현황을 조회할 수 있다

KJungW committed:
test(search_record): 제적 위험도 계산

테스트 내용
- 닉네임을 통해 해당 크루의 제적 위험도를 계산할 수 있다

KJungW committed:
feat(search_record): 제적 휘험도 계산 (구현)

구현 내용
- 닉네임을 통해 해당 크루의 제적 위험도를 계산할 수 있다

KJungW committed:
test(search_record): 등록되지 않는 크루 처리 (실패 테스트 작성)

구현내용
- 닉네임이 등록되지 않은 경우 예외 메세지를 출력한다

KJungW committed:
feat(search_records): 등록되지 않은 크루 처리 (구현)

구현내용
- 닉네임이 등록되지 않은 경우 예외 메세지를 출력한다

KJungW committed:
test(risk): 제적 위험자 조회 (실패 테스트 구현)

테스트 내용
- 크루마다의 출석, 지각, 결석 횟수를 계산할 수 있다

KJungW committed:
feat(risk): 제적 위험자 조회 (구현)

구현내용
- 제적 위험이 있는 크루의 출석 상태와 제적 위험도를 구할 수 있다

KJungW committed:
test(risk): 제적 위험자 정렬 (실패 테스트 구현)

테스트내용
- 제적 위험이 있는 크루의 출석 상태와 제적 위험도를 구할 수 있다
- 제적 위험자 첫번째로 제적 위험도의 내림차순으로 정렬된다
- 제적 위험자 두번째로 결석 지각의 내림차순으로 정렬된다

KJungW committed:
feat(risk): 제적 위험자 정렬 (구현)

구현내용
- 제적 위험자 첫번째로 제적 위험도의 내림차순으로 정렬된다
- 제적 위험자 두번째로 결석 지각의 내림차순으로 정렬된다
- 제적 위험자 세번째로 닉네임의 내림차순으로 정렬된다

KJungW committed:
feat(menu): 메뉴 입력 (구현)

구현내용
- 메뉴 기능을 선택하는 커맨드를 입력받는다
- 선택된 커맨드가 유효하지 않은 경우 예외를 발생시킨다.

KJungW committed:
feat(menu): 메뉴별 기능 실행 (구현)

구현내용
- 선택된 메뉴에 해당하는 기능을 실행한다
- 프로그램은 사용자가 종료할 때까지 종료되지 않으며,
  기능 수행 후 초기화면으로 돌아간다

KJungW committed:
refactor(application): 패키지 정리

작업내용
- 잘못 배치된 파일과 패키지 정리

KJungW committed:
feat(initilze): 초기 데이터 세팅 (구현)

구현내용
- 초기 데이터가 담긴 파일을 통해 초기 데이터 세팅

KJungW committed:
feat(output): 기능 수행 결과 출력 (구현)

구현내용
- 메뉴 기능들을 수행한고 각각의 결과를
  출력하는 기능 구현

KJungW committed:
fix(risk): 위험도 계산 버그 수정

수정내용
- 제적 위험도를 계산하는 계산식을 수정

KJungW committed:
fix(input): 메뉴 내용 수정

수정내용
- 메뉴의 요일이 현재 요일로 나오도록 수정

KJungW committed:
refactor(output): 출력 메세지 상수화

구현내용
- 출력 메세지를 별도의 Enum으로 분리

KJungW committed:
refactor(application): 패키지 정리

작업내용
- 잘못된 위치에 배치된 패키지를 정상 위치에 배치

KJungW committed:
refactor(add_record): 출석 기록 저장소 분리

작업내용
- 출석 시스템 도메인에서 출석 기록 저장소를 분리

KJungW committed:
test(chekcer): checker 관련 도메인 테스트

테스트내용
- AttendanceChecker, AttendanceType, CampusTime,
  HolidayChekcer 관련 테스트 추가

KJungW committed:
test(crew): crew 패키지 관련 테스트

테스트 내용
- Crew, CrewStorage 관련 테스트 추가

KJungW committed:
test(risk): risk 패키지 관련 테스트

테스트 내용
- RiskType 테스트 추가

KJungW committed:
test(record): record 패키지 관련 테스트

테스트 내용
- AttendanceRecord, AttendanceRecordStorage에 대한
  테스트 추가

KJungW committed:
feat(utility): 날짜 시간 유틸리티

구현내용
- 날짜와 시간을 파싱하고 포맷팅하는 유틸리티 생성

KJungW committed:
test(input): 입력값 검증 (실패 테스트 작성)

테스트 내용
- 비어있는 값이 입력되었는지 검증할 수 있다
- 숫자 형식 아닌 데이터가 입력되었는지 검증할 수 있다

KJungW committed:
test(input): 입력값 검증 (구현)

구현내용
- 입력값에 대한 기본 검증 구현
- 검증을 입력뷰에 적용

KJungW committed:
test(input): 입력값 전처리 (실패 테스트 작성)

테스트 내용
- 입력값의 양옆 공백을 제거할 수 있다

KJungW committed:
feat(input): 입력값 전처리 (구현)

구현내용
- 입력값의 사이드 공백을 제거하는 기능 구현
- 구현한 기능을 입력뷰에 적용

KJungW committed:
feat(record): 입력 닉네임 검증 (구현)

구현내용
- 입력된 닉네임이 등록되지 않은 크루일 경우
  예외를 발생시킨다

KJungW committed:
test(input): 날짜 유효성 검증 (실패 테스트 작성)

테스트 내용
- 달에 포한된 날짜가 입력되었는지 검증할 수 있다

KJungW committed:
feat(input): 날짜 유효성 검증 (구현)

구현내용
- 입력된 날짜가 유효한 날짜인지 검증
- 구현된 검증을 실제로 적용

KJungW committed:
refactor(exception): 커스텀 예외 적용

구현 내용
- 커스텀 예외를 생성하고 적용

KJungW committed:
refactor(output): 출력 형식 변경

구현 내용
- 제적 위험도가 없는 크루는 출력 조회시 제적 휘험도를
  출력하지 않도록 수정

KJungW committed:
refactor(menu): 메뉴 커맨드 필드 형식 변경

구현내용
- 필드를 final로 수정

KJungW committed:
refactor(record): 메서드 분리

구현내용
- AttendanceType의 길이가 긴 메서드를 분리

KJungW committed:
style(application): 코드 정리

KJungW committed:
refactor(utility): 메서드명 수정

수정내용
- 메서드명에 포함된 오타 수정

KJungW committed:
style(application): 코드 정리

seovalue commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,114 +1,98 @@
+    }
+
+    public void run() {
+        while (true) {
```
요기서의 indent는 불필요하게 while을 빼지 않기 위해서 indent2를 허용해주신 걸까요?

KJungW replied:
넵! 맞습니다. 처음에는 기능별로 while문을 따로 빼서 관리했는데 이 경우 휴일에 출석하려고 할때 “휴일 출석 시도→휴일이라 출석 예외 발생→while문으로 다시 휴일 출석 시도”의 루프에 빠지게 되었습니다. 그래서 이를 방지하고자 기능 수행 중 예외가 발생하면 해당 기능이 아닌 메뉴 입력부터 다시 시작하도록 하기위해 모든 기능을 하나의 while로 묶어버렸고, 결과적으로 indent가 2가 되었습니다.

seovalue commented on src/main/java/attendance/domain/AttendanceSystem.java:
```diff
@@ -1,105 +1,94 @@
-        AttendanceType attendanceType = CampusSchedule.calculateAttendanceType(arrivalDateTime);
-        AttendanceRecord newRecord = new AttendanceRecord(nickname, arrivalDateTime, attendanceType);
+    public AttendanceRecord addAttendanceRecord(String nickname, LocalDateTime arrivalDateTime) {
+        crewStorage.validateIsNotContained(nickname);
```
여기서 해주는 검증과 Controller에서 not registered일 때 해주는 검증은 어떤 차이가 있나요?

KJungW replied:
컨트롤러에서 하는 검증은 “등록되지 않은 닉네입 입력값에 대한 즉각적인 예외처리를 위해서” 라는 의도를 담았고, 이를 통해 잘못된 닉네임이 입력된 경우 다른 추가적인 입력을 진행하기 전에 사용자에게 즉각적으로 잘못된 입력임을 알릴 수 있었습니다.
반면에 AttendanceSystem에서 해주는 검증은 “주요 도메인 로직을 실행하기 전에 크리티컬한 오류를 방지하기 위한 검증”이라는 의도를 담았습니다. 이를 통해 주요 도메인 로직을 실행하기 전 혹시나 발생할 수 있는 오류를 방지할 수 있었습니다.
해당 검증이 중복으로 존재하는 것이 맞을까 고민해보았지만 두 검증 모두 자신만의 역할이 있기 때문에 중복으로 존재해도 좋지않을까 하는 생각이 듭니다!

seovalue replied:
굿굿 저도 중복 검증하는 것은 좋아보여요.
다만 만약 CrewStorage가 DB 였다면, 같은 검증을 위해 DB 쿼리를 두번 호출하게 될텐데 이 부분에 대해서는 어떻게 생각하시는지 궁금해요

KJungW replied:
우선 성능상에 문제가 없다면 이대로도 괜찮을 것 같습니다. 만약 성능상에 문제가 생겨 리펙토링이 필요하다면 각각의 기능을 하나의 트랜젝션으로 생각하고, 트랜젝션 내에서만 유효한 캐시를 만들어 활용하는 방안으로 쿼리의 수를 줄여볼 것 같습니다.

seovalue commented on src/main/java/attendance/domain/checker/AttendanceChecker.java:
```diff
@@ -0,0 +1,38 @@
+        this.holidayChecker = holidayChecker;
+    }
+
+    public AttendanceType checkAttendance(LocalDateTime arrivalDatetime) {
```
checkAttendance는 정확히 어떤걸 체크하는걸까요?

AttendanceChecker.checkAttendacne 만을 봤을 때에는 구현을 유추하기가 어려울 것 같아요

KJungW replied:
확실히 제가 생각하지 못한 부분인 것 같습니다. 출석 타입(AttendanceType 열거형)을 체크하는 클래스라는 의미에서 클래스명을 AttendanceTypeChecker로 하고, 메서드명을 check()나 checkType() 정도로 해주면 의미전달이 좀더 명확해질 것 같습니다. 개인적으로 AttendanceTypeChecker에 다른 check 메서드를 만들지는 않을 것 같아 check()라는 메서드명을 선택해 리펙토링을 진행해보겠습니다!

seovalue commented on src/main/java/attendance/domain/initializer/AttendanceSystemInitializer.java:
```diff
@@ -0,0 +1,40 @@
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+
+public class AttendanceSystemInitializer {
```
👍

KJungW replied:
감사합니다!

seovalue commented on src/main/java/attendance/domain/crew/CrewStorage.java:
```diff
@@ -8,23 +9,24 @@
-            throw new IllegalArgumentException(ExceptionMessage.NOT_FOUND_CREW.getContent());
-        }
+    public boolean checkIsNotContained(String nickname) {
+        return crews.stream().noneMatch(crew -> crew.isSameNickname(nickname));
```
check할 때마다 stream으로 일일이 조회하게 되는데 내부에서는 map으로 관리하는 건 어떤가요?
https://github.com/woowacourse/java-attendance/pull/15#discussion_r1966650648
여기서도 한번 이야기 나눴던거긴 한데 CrewStorage 안에서도 map으로 관리되어도 될 것 같아서요~ 어떻게 생각해요?

KJungW replied:
말씀해주신 Map<크루이름, Crew>면 크루의 중복을 허용하지 않는 문제와 중복된 조회 문제를 모두 간편하게 해결할 수 있을 것 같습니다. Crew에 필드가 하나라 Map을 적용할 수 있다는 생각을 해본 적 없었는데 이번 피드백을 통해 다음부터는 Map을 활용한 인덱싱을 좀더 다양한 상황에 적용해볼 수 있을 것 같습니다!

seovalue commented on src/main/java/attendance/domain/record/AttendanceRecordStorage.java:
```diff
@@ -1,54 +1,68 @@
 public class AttendanceRecordStorage {
 
-    private final Map<String, AttendanceRecords> attendanceRecords = new HashMap<>();
+    private final List<AttendanceRecord> records = new ArrayList<>();
```
오버스펙일 수도 있는데 랜디 의견 궁금해서 하나 남겨봐요!
    private final Map<AttendanceKey, AttendanceRecord> records = new HashMap<>();

 private static class AttendanceKey {
        private final String nickname;
        private final LocalDate date;
}

뭔가 위처럼 관리되는 건 어떤 것 같아요?
항상 찾을 때 nickname과 date 기반으로 조회하는 것 같아서, 이게 key가 되는 형태가 되면 전반적인 성능을 O(1)으로 할 수 있지 않을까 싶어서요!

KJungW replied:
2단계 미션 요구사항인 “TDD를 통해 요구사항에 맞는 최소한의 설계를 해보자”를 실천하기 위해 리스트에서 더 리펙토링을 진행하지는 않았지만 이전부터 리펙토링이 필요할 것 같다고 생각한 부분이었습니다. 2단계 미션을 만들고 제출하는 과정을 통해 해당 요구사항으로 배워야할 점을 어느정도 학습했다고 생각하기 때문에 현재 단계에서는 오버스펙이라도 성능향상을 위한 리펙토링을 적용해도 좋을 것 같다고 생각합니다. 
저는 원래 크루의 닉네임으로 인덱싱을 하는 방향을 고려중이었습니다. 하지만 조앤의 피드백을 보고 출석 기록 하나하나에 대해 인덱싱하는 방향을 새롭게 알게되었고 확실히 기존의 방식보다 속도면에서 큰이점이 있을 것 같습니다. Map을 활용한 인덱싱에 대해 보다 시야가 넓어지는 것 같습니다!

seovalue commented on src/main/java/attendance/domain/crew/Crew.java:
```diff
@@ -2,20 +2,20 @@
-    public Crew(String name) {
-        this.name = removeSpaceInName(name);
+    public Crew(String nickname) {
+        this.nickname = nickname;
```
크루명에 대한 validation은 따로 필요 없을까요?

KJungW replied:
요구사항에 명시되어 있지는 않아 신경쓰지 못한 부분이었는데 확실히 시스템 상으로 검증이 필요하다는 생각이 듭니다! 이번에 리펙토링에서 닉네임이 null, ‘’, ‘ ‘인 경우 검증 처리를 추가해보겠습니다. 그리고 양 사이드 공백을 제거하는 전처리 코드도 추가해보겠습니다!

seovalue commented on src/test/java/attendance/domain/checker/CampusTimeTest.java:
```diff
@@ -0,0 +1,55 @@
+                .isEqualTo(inInCampusTime);
+    }
+
+    static Stream<Arguments> 캠퍼스_운영시간인지_체크할_수_있다() {
```
와우 테스트를 진짜 꼼꼼히 작성해주셨네요 👍

KJungW replied:
감사합니다!

seovalue commented on src/test/java/attendance/domain/record/AttendanceRecordStorageTest.java:
```diff
@@ -1,126 +1,109 @@
-
-        assertThat(recordStorage.findUnmodifiedRecordsByNickname("쿠키", 2024, Month.DECEMBER)).hasSize(3);
+    void 결석_포함_출석_조회_결석을_포함해서_출석을_조회한다() {
+        recordStorage.add(RECORD);
+        AttendanceRecord attendanceRecord = recordStorage.findWithAbsenceRecord(NICKNAME, ARRIVAL_DATE);
+        assertThat(attendanceRecord.getAttendanceType()).isEqualTo(ATTENDANCE_TYPE);
+
+        recordStorage.remove(NICKNAME, ARRIVAL_DATE);
+        AttendanceRecord absenceRecord = recordStorage.findWithAbsenceRecord(NICKNAME, ARRIVAL_DATE);
+        assertThat(absenceRecord.getAttendanceType()).isEqualTo(AttendanceType.ABSENCE);
```
이 코드를 읽기 위해서는 상수화된 값을 확인하기위해 계속 위로 이동해야하는 것 같아요
테스트 안에서 상수를 잘 관리하려면 어떻게 해보는 것이 가독성을 높일 수 있을까요??

KJungW replied:
이번 금요일 수업에서 Test Fixture의 활용에 대해 새롭게 알게되었습니다. 저는 Test Fixture가 테스트용 객체를 생성하는 용도라고만 생각했는데 테스트 상수도 모아서 관리할 수 있다는 것을 깨닫게 되었습니다. 그래서 반복적으로 사용되는 테스트 상수는 Test Fixture로 만들어 관리하고, 나머지 테스트 클래스에서 개별적으로 사용되는 상수는 테스트 클래스 자체에서 관리하는 방안으로 리펙토링을 진행해보겠습니다!

seovalue reviewed (changes requested):
랜디! 늦어져서 죄송해요ㅠㅠ
우선 커밋로그 확인해봤을 때 전체적으로 TDD로 너무 잘 구현해주신 것 같아요 💯 
소소한 코멘트 몇가지 남겨뒀는데 확인 후 리뷰요청 주세요! 즐거운 금요일 화이팅~

seovalue commented:
> TDD 진행방식에 질문이 있습니다.
> 
> 이번에 가장 큰 도메인인 AttendanceSystem을 기준으로 설계없이 TDD를 진행했습니다. 진행하면서 AttendanceSystem의 내부 도메인들은 언제든지 바뀔 수 있다는 생각에 AttendanceSystem에 대한 테스트만을 작성하며 TDD로 기능구현을 진행해나갔고, 어느정도 기능구현이 완료되었을 때 AttendanceSystem 내부 도메인들을 확정시키고 해당 내부 도메인들에 대한 테스트를 작성했습니다. TDD는 항상 실패 테스트 먼저라고 배우기는 했는데 내부 도메인들이나 설계가 확정되지 않은 상태에서 내부 도메인에 대한 테스트를 작성하는 것은 비효율적이라고 생각해 이러한 방식으로 진행했고, 이전 TDD를 진행할때보다 TDD의 이점이 잘 느껴졌습니다. 현재 진행방식이 바람직한 방식인지 조앤의 의견이 궁금합니다!

저는 보통 개발을 시작하기 전에 설계를 진행하는 편이고, 설계가 완료되었다면 그 안에서 가장 작은 도메인부터 채워나가는 것 같아요!!


> 커스텀 예외 생성 기준이 궁금합니다.
> 커스텀 예외를 통해 예외의 범위를 나누라는 것을 배우고 이번 미션에서는 커스텀 예외를 만들어 적용했습니다. 항상 할때마다 커스텀 예외를 어떤 기준으로 나눠야할지 고민이 많이 되는데 조앤은 어떤 방식으로 커스텀 예외를 생성하시는지 궁금합니다!

랜디가 작업해주신 것은 Message는 하나로 관리하되 객체가 구분된 형태인 것 같아요. 
또 다른 방법으로는 exception 객체와 message를 1:1로 매핑시키는거예요. 즉 AlreadyAttendanceException()과 같이 객체 자체를 상세히 두는거죠. 

또 생각나는건 위의 랜디의 방법에서 예외 객체를 하나로 관리하는 방법도 있을 것 같아요. 즉 하나의 Exception에 커스텀한 메시지로 구분하는 거죠.

사실 컨벤션에 따라서 진행하는 거라, 이건 협업의 관점이긴 한데 랜디는 여러가지 방법들 중 지금의 구현을 택해주신 이유가 궁금하네요~~ 그이후 저도 제 의견 드려볼게요!!

KJungW committed:
refactor(record): 불분명한 클래스명과 메서드명 수정

작업내용
- AttendanceChecker를 AttendanceTypeChecker로 수정
- checkAttendance()를 check()로 수정

KJungW committed:
refactor(crew): 인덱싱 적용

구현내용
- map을 활용해 crew들을 인덱싱해서 관리

KJungW committed:
test(crew): 크루 생성 검증 (실패 테스트 작성)

테스트 내용
- 닉네임이 공백일 경우 크루를 생성할 수 없다

KJungW committed:
feat(crew): 크루 생성 검증

구현내용
- 닉네임이 공백일 경우 크루를 생성할 수 없다

KJungW committed:
test(crew): 크루 닉네임 전처리 (실패 테스트 작성)

테스트내용
- 크루를 생성할 때 입력된 닉네임의 양사이드 공백을 제거한다

KJungW committed:
feat(crew): 크루 닉네임 전처리 (구현)

구현내용
- 크루를 생성할 때 입력된 닉네임의 양사이드 공백을 제거한다

KJungW committed:
refactor(record): 출석 기록 인덱싱

구현내용
- 출석 기록을 map으로 저장하여 인덱싱 적용

KJungW committed:
test(fixture): 상수 관리 fixture 생성 및 적용

구현 내용
- 자주 사용되는 테스트 상수를 모아 관리하는
  fixture를 생성 및 적용

KJungW committed:
test(record): 변수명 수정

구현내용
- 가독성을 위해 상수명과 변수명 수정

KJungW committed:
test(record): 불필요한 상수 제거

구현내용
- 가독성을 해치는 불필요한 상수 제거

KJungW commented:
exception 객체와 message를 1:1로 매핑시키는 방식은 새롭게 알게되었습니다! 이 방식으로 커스텀 예외를 구현하면 예외상황을 보다 세밀하게 표현이 가능할 것 같습니다. 나중에 예외 상황을 세밀하게 구체화 해야하는 경우 새롭게 알게된 일대일 매핑방식을 사용해보겠습니다.

그리고 이번 프로젝트에서는 현재 예외처리 방식을 선택한 이유를 말씀드리겠습니다. 우선 키워드 단위로 커스텀 예외 클래스를 구성해 예외상황을 보다 구체적으로 표현해보려고 했습니다. 그리고 예외메세지의 경우 시스템의 크기가 작기 때문에 예외메세지를  한 곳에 모아서 보다 편리하게 관리할 수 있도록 구성했습니다. 현재  방식에 대해 조앤이 짚어주실 부분이 있다면 경청하겠습니다!

그리고 조앤이 말씀해주신 대로 커스텀 예외처리 방식은 컨밴션 즉 선택의 문제라는 생각이 듭니다. 시스템 특성에 맞게 적절히 예외처리 방식을 선택하는 방법을 앞으로 고민해보겠습니다!

seovalue commented on src/main/java/attendance/domain/AttendanceSystem.java:
```diff
@@ -66,7 +67,7 @@ public AttendanceState calculateAttendanceStateInMonth(String nickname, LocalDat
 
     public List<AttendanceState> findRiskCrew(LocalDate today) {
-        List<String> allNicknames = crewStorage.findAllNicknames();
+        Set<String> allNicknames = crewStorage.findAllNicknames();
```
👍

KJungW replied:
감사합니다!

seovalue commented on src/main/java/attendance/domain/AttendanceSystem.java:
```diff
@@ -89,6 +90,6 @@ private int calculateMaxAttendanceCountUntilTodayInMonth(LocalDate today) {
 
     private List<LocalDate> calculateNotHoliday(int year, Month month) {
-        return attendanceChecker.calculateNotHolidayInMonth(year, month);
+        return attendanceTypeChecker.calculateNotHolidayInMonth(year, month);
```
랜디 이 역할을 attendanceTypeChecker 내부에서 하는 이유가 뭐예요?

KJungW replied:
AttendanceTypeChecker로 클래스명을 구체화하면서 해당 메서드가 AttendanceTypeChecker에 어울리지 않는 메서드가 된 것 같습니다. 리펙토링 방안을 생각해보면 HolidayChecker에 해당 메서드를 배치하고 AttendanceSystem에서 HolidadyChecker를 주입받아 해당 메서드를 사용하는 방식을 고려할 것 같습니다!

seovalue commented on src/main/java/attendance/domain/crew/Crew.java:
```diff
@@ -1,13 +1,16 @@
 
     public Crew(String nickname) {
-        this.nickname = nickname;
+        validateBlankNickname(nickname);
```
👍 리뷰 반영해주셔서 감사합니다
나중에는 이게 CrewName이 될 수도 있겠네요!

KJungW replied:
확실히 크루명을 원시값 포장 객체로 관리하면 관련 로직을 묶어 관리할 수 있어 좋을 것 같습니다. Crew 도메인에서도 nickname필드를 CrewName관리하면 중복코드없이 크루의 닉네임을 보다 객체지향적으로 관리할 수 있을 것 같습니다!

seovalue commented on src/main/java/attendance/domain/crew/CrewStorage.java:
```diff
@@ -2,22 +2,22 @@
 
-    private final Set<Crew> crews = new HashSet<>();
-    
+    private final Map<String, Crew> crews = new HashMap<>();
```
여기도 CrewName을 사용하게 된다면 좀 더 명확하게 관리될 수 있을 것 같아요 (String > CrewName)

KJungW replied:
CrewName이라는 원시값 포장객체를 활용하게 되면 관련 로직을 묶어서 관리할 수 있다는 이점 외에 코드의 가독성을 높일 수 있다는 점 새롭게 배웠습니다!

seovalue commented on src/main/java/attendance/domain/record/AttendanceRecordStorage.java:
```diff
@@ -65,4 +69,12 @@ private AttendanceRecord makeAbsenceRecord(String nickname, LocalDate arrivalDat
         return new AttendanceRecord(nickname, dateTime, AttendanceType.ABSENCE);
     }
+
+    private List<AttendanceRecordIdentifier> findIdentifierInMonth(String nickname, int year, Month month) {
+        Set<AttendanceRecordIdentifier> allIdentifiers = records.keySet();
```
외부로 전달할 때에는 Identifier 말고 별도의 DTO로 이름을 줘도 괜찮을 것 같은 느낌이 드네요!

KJungW replied:
Identifier가 불변객체이기도 하고 필드 개수가 두 개 뿐이라 DTO가 아닌 Identifier가 외부로 전달되어도 문제는 없을 것 같습니다. 하지만 가독성 측면에서 생각해보면 외부로 전달되는 데이터에 대한 보다 의미있는 설명을 위해 DTO를 새롭게 구성하는 것도 좋은 리펙토링 방향인 것 같습니다!

seovalue commented on src/test/java/attendance/domain/crew/CrewTest.java:
```diff
@@ -1,12 +1,35 @@
+    @DisplayName("닉네임이 공백일 경우 크루를 생성할 수 없다")
+    @ParameterizedTest
+    @NullAndEmptySource
+    void 닉네임이_공백일_경우_크루를_생성할_수_없다(String input) {
```
👍 테스트도 함께 추가해주셨군요!

seovalue commented on src/test/java/attendance/fixture/CrewNicknameFixture.java:
```diff
@@ -0,0 +1,7 @@
+
+public class CrewNicknameFixture {
+
+    public static final String VALID_CREW_NICKNAME = "쿠키";
```
Fixture를 사용할 때엔 영어로 표현하기 길어지는 경우 한국어를 사용하기도 해요! 나중에 한번 해보셔도 좋을 듯ㅎㅎ

KJungW replied:
확실히 영어이름이라 상수명이 길어져서 곤란했었습니다! 테스트 데이터만을 위한 상수이니 한글로 지어서 보다 짧고 명확하게 의미전달을 하는 편이 가독성과 유지보수 측면에서 좋을 것 같습니다. 새로운 지식 감사합니다!

seovalue commented on src/test/java/attendance/fixture/DateFixture.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.time.LocalDate;
+
+public class DateFixture {
```
Fixture를 사용해보니 어떠셨는지 궁금해요~~

KJungW replied:
테스트 클래스 상단에 테스트 상수가 엄청 많다는 점, 비슷한 상수를 테스트 클래스마다 별도로 관리하는 점 모두 불편했었는데 Ftxture를 통해 불편했던 점을 해결해서 속이 시원했습니다! 다음부터도 Fixture를 자주 사용하게 될 것 같습니다. 
보완해야할 점으로는 Fixture에서 상수만 관리하는 것이 아니라 공통 테스트 데이터를 생성하는 본래의 역할도 수행하도록 해서 테스트의 가독성을 좀더 높여보고 싶습니다!

seovalue reviewed (approved):
랜디! 1단계 첫번째 제출했을 때에 비해서 지금 코드를 보았을 때 훨씬 깔끔해지고 가독성도 좋아졌다고 생각해요. 미션도 빠르게 수행해주시고 리뷰도 적절히 반영해주셔서 감사합니다!
소소한 코멘트만 몇가지 남겨뒀어요. 다음 단계 진행하시기 전에 한번 확인해주세요~
고생 많으셨어요!! 👍
